src/tracker-component.js

```diff
@@ -0,0 +1,102 @@
+
+type Product = {|
+    id : string,
+    url : string,
+    sku? : string,
+    name? : string,
+    description? : string,
+    imgUrl? : string,
+    otherImages? : $ReadOnlyArray<string>
+|};
+
+type Cart = {|
+    cartId? : string,
+    items : $ReadOnlyArray<Product>,
+    emailCampaignId? : string,
+    price? : number,
+    currencyCode? : string,
+    keywords? : $ReadOnlyArray<string>
+|};
+
+type RemoveCart = {|
+    cartId? : string,
+    items : $ReadOnlyArray<{ id : string }>
+|};
+
+type ParamsToBeaconUrl = ({
+    trackingType : TrackingType,
+    data : Cart | RemoveCart | { cartId : string } | { pageUrl : string }
+}) => string;
+
+type Config = {|
+    user : {
+        id : string,
+        email? : string, // mandatory if unbranded cart recovery
+        name? : string
+    },
+    property? : {
+        id : string
+    },
+    paramsToBeaconUrl? : ParamsToBeaconUrl,
+    env? : Environment
+|};
+
+const track = <T>(config : Config, trackingType : TrackingType, trackingData : T) : Promise<void> => {
+    const encodeData = data => encodeURIComponent(btoa(JSON.stringify(data)));
+
+    return new Promise((resolve, reject) => {
+        const img = document.createElement('img');
+        img.style.display = 'none';
+
+        const data = {
+            ...trackingData,
+            user:       config.user,
+            property:   config.property,
+            trackingType,
+            clientId:   getClientID(),
+            merchantId: getMerchantID()
+        };
+
+        if (config.paramsToBeaconUrl) {
+            img.src = config.paramsToBeaconUrl({ trackingType, data });
+        } else {
+            const envToUrl = {
+                stage:      'https://www.targetingnodeweb19125146982616.qa.paypal.com',
+                sandbox:    'https://sandbox.paypal.com',
+                production: 'https://www.paypal.com'
+            };
+            img.src = `${ envToUrl[config.env || 'production'] }/targeting/track/${ trackingType }?data=${ encodeData(data) }`;
+        }
+
+        img.addEventListener('load', () => resolve());
+        img.addEventListener('error', reject);
+        if (document.body) {
+            document.body.appendChild(img);
+        }
+    });
+};
+
+const trackCartEvent = <T>(config : Config, cartEventType : CartEventType, trackingData : T) : Promise<void> =>
+    track(config, 'cartEvent', { ...trackingData, cartEventType });
+
+export const Tracker = (config : Config) => ({
+    view:           (data : { pageUrl : string }) => track(config, 'view', data),
+    addToCart:      (data : Cart) => trackCartEvent(config, 'addToCart', data),
+    setCart:        (data : Cart) => trackCartEvent(config, 'setCart', data),
+    removeFromCart: (data : RemoveCart) => trackCartEvent(config, 'removeFromCart', data),
+    purchase:       (data : { cartId : string }) => track(config, 'purchase', data),
+    setUser:        (data : { user : { id : string, name? : string, email? : string } }) => {
```

Comment:
why do we have this method `setUser`?

Reply:
It's possible that we don't know the user when we start tracking their cart events, but later they might log in and we'll know who they are, so we can use `setUser`.